### DISCUSSION TITLE
When installing on Apache web server there's no CSS?

### DISCUSSION THREAD

**Discussion by msharp1** [datenstrom/community#120] [Ask a question] [answered]
Hi,
I've just unzipped the zip file, and [here is what I get](http://mhash.fr/yellow/) ?
There is a 404 on http://mhash.fr/yellow/media/themes/flatsite.css (there is no such file, indeed).

**Comment by markseuffert** [comment]
Hi, which web server are you running? Looks like you need to update your server configuration.

**Comment by msharp1** [comment]
Thanks for the reply, here it is http://mhash.fr/yellow/info.php
Olympe is a free hosting solution but I still can ask the support is something needs to be changed.

**Comment by markseuffert** [comment]
Sorry, I don't know about Olympe web hosting. I tried to register an account but they temporarily closed.

Please check if you have uploaded the `.htaccess` file and [contact Olympe support](https://www.olympe.in/about/contact) if it still doesn't work.

**Comment by msharp1** [comment]
Thanks markseu, I think there is a problem with the .htaccess file indeed. (but I didn't spot the exact issue). I'll contact the support.

**Comment by msharp1** [comment]
Today I downloaded the new Yellow version (https://github.com/datenstrom/yellow/archive/master.zip), unzipped and upload all files with the same results.

@markseu , what could the the reason why yellow is trying to search a CSS file at /media/themes/flatsite.css , which file can't be found there ?
I've opened an issue at https://community.olympe.in/t/yellow-cms-ne-fonctionne-pas-correctement/10138

**Comment by msharp1** [comment]
Shouldn't yellow look for system/themes/flatsite.css instead of media/themes/flatsite.css ?

**Comment by msharp1** [comment]
it seems core.php is looking for $themeLocation dir, when the css file is in $themeDir dir
What's the difference between xxxLocation and xxxDir config parameters ?

**Comment by markseuffert** [comment, marked as answer]
Yellow requires a web server with rewrite module. It looks like the rewrite module is not properly working on your web server. Once this is configured, your Yellow installation should work. 

To answer your question, the location for the `/media/themes/flatsite.css` is fine, usually `yellow.php` takes care of this. That's why the rewrite module is needed, it tells Yellow to handle such requests. Sorry, I don't know how to help without an account at Olypme. I contacted them via Twitter and they don't know when registration will be reopened again. Anyone has a test account I can use?

**Comment by msharp1** [comment]
@markseu, I've just contacted you using the contact form on your website, so we can get in touch to find a solution to get you a test account. Thanks !

**Comment by markseuffert** [comment]
Here's what I found out: The web hoster uses Apache/2.4.10 and you need to modify your `.htaccess` file. Once you added the subfolder as [RewriteBase](https://httpd.apache.org/docs/2.4/mod/mod_rewrite.html#rewritebase) the Yellow installation works. What puzzles me is that I am running several Yellow installations in a subfolder... and it works without modification.

Here's an example (updated):

``` apache
<IfModule mod_rewrite.c>
RewriteEngine on
RewriteBase /yellow/
DirectoryIndex index.html yellow.php
RewriteRule ^(cache|content|system)/ error [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^ yellow.php [L]
</IfModule>
```

I don't understand why this is necessary, but it does the trick. Hope this helps. :smile:

**Comment by msharp1** [comment]
Hi !
Thank you very much for your help.
Indeed this is very strange but it is likely possible that Olympe apache's configuration is specific (ex: hosted files are not in DocumentRoot ?), that so it requires a RewriteBase directive.
But as it might happen to other people, here is what I suggest : why not adding this RewriteBase directive commented in the .htaccess file provided with yellow ? With something saying : "This might be necessary with specific apache configuration". Something like that.
I'm not sure to exactly understand the issue so I wont comment more than that :D

**Comment by markseuffert** [comment]
Thanks for your feedback. We can add it later, when it happens on more than on web hoster. I understand this isn't ideal for installing software at Olympe. Let's keep the issue open for a while and see if more people are affected.

**Comment by msharp1** [comment]
There could be people that silently abandon because of this... but its your décision : do whatever you think its right. Cheers!

Le 31 octobre 2015 22:45:03 CET, Mark Seuffert notifications@github.com a écrit :

> Thanks for your feedback. We can add it later, when it happens on more
> than on web hoster. I understand this isn't ideal for installing
> software at Olympe, Glad to hear it works now.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/datenstrom/yellow/issues/120#issuecomment-152775101

## 

Envoyé de mon téléphone Android avec K-9 Mail. Excusez la brièveté.

**Comment by msharp1** [comment]
Hi @markseu ,

I tried installing yellow in a $USER/public_html of a fresh Ubuntu box (15.10) with apache2 and php5. Enabled 'userdir' mod. I can't get it to work, meaning I get unstyled yellow home page ie same symptom I've initially reported in this issue. Also tried various RewriteBase combinations in the .htaccess with no success.

As a side note, on the same box I tried to install yellow in a regular folder (ie /srv/web) and I had to specify 'AllowOverride Indexes' (in apache2.conf) to make it work - because yellow .htaccess uses DirectoryIndex directive. I think this should be specified somewhere, not every web environment has this directive enabled.

So back to the first issue : I would appreciate any suggestion.

**Comment by markseuffert** [comment]
What's in your Apache configuration for `<Directory /home/*/public_html>`?

**Comment by msharp1** [comment]
Hi Mark,
Pretty much the defaults settings which resides in `mods-enabled/userdir.conf`, just added a LogLevel directive : 

```
<IfModule mod_userdir.c>
        UserDir public_html
        UserDir disabled root
        LogLevel alert rewrite:trace3
        <Directory /home/*/public_html>
                AllowOverride FileInfo AuthConfig Limit Indexes
                Options MultiViews Indexes SymLinksIfOwnerMatch IncludesNoExec
                <Limit GET POST OPTIONS>
                        Require all granted
                </Limit>
                <LimitExcept GET POST OPTIONS>
                        Require all denied
                </LimitExcept>
        </Directory>
</IfModule>
```

**Comment by markseuffert** [comment]
I would use `AllowOverride All` for the directory containing Yellow. This should do the trick.

**Comment by msharp1** [comment]
Hi Mark, already tried that - switching the default `AllowOverride FileInfo AuthConfig Limit Indexes` to `AllowOverride All` - with no more success.

**Comment by msharp1** [comment]
Here is the corresponding rewriterule trace : http://pastebin.com/029kM3sb
(I don't see anything wrong with it, even css file is passed to yellow ? `rewrite 'media/themes/flatsite.css' -> 'yellow.php'`)

**Comment by msharp1** [comment]
So I have found that this time, considering yellow installed in `$USER/public_html/yellow` that RewriteBase should be set to `RewriteBase /~user/yellow`

I found the solution but I'm not sure to understand the reason and how to debug. Could you give a hint Mark ?

**Comment by msharp1** [comment]
Ok I think I get it. if `RewriteBase` is not specified, mod_rewrite use the physical path (ie `/home/$USER/public_html/yellow`) as the default RewriteBase URI prefix; so you basically get a 404 because `http://server/home/$USER/public_html/yellow` leads to nothing. So you have to tell explicitely mod_rewrite that the URI prefix is `~user/yellow`
If not exactly this, very close to :)

So basically it means that RewriteBase is mandatory when mod_userdir is used with yellow. Again, this could be specified somewhere in the documentation :)

**Comment by markseuffert** [comment]
Thank you very much for your investigation.

Now who's gonna document this for Yellow? Anyone wants to help with a trouble shooting section? :smile:

**Comment by markseuffert** [comment]
I updated web server configurations in the [troubleshooting](https://datenstrom.se/yellow/help/troubleshooting).

From my point of view the Yellow documentation should provide examples for the most common web hostings. I don't think we should handle all possible Apache configurations and modules. I think it would be nice if Yellow could automatically work with most web servers (and PHP versions out there) and detect if something is wrong with the web server. What do you think? If you want to help, please update the Yellow documentation and create new issues for installation problems.

**Comment by glasen** [comment]
Hi,
i'm just running into basically the same problem. I've copied Yellow to my university file space and now i'm trying to get it running. I just need the correct "RewriteBase".

Here is my URL for the installation:

https://homepages.thm.de/~sgls22/

When i'm trying to open the "about"-link there is a 404-error:

> The requested URL /var/chroot/home/sgls22/public_html/yellow.php was not found on this server.

I'm currently using the normal htaccess-file provided with the Yellow-package.

**Comment by markseuffert** [comment]
Can you add a RewriteBase with your username to the `.htaccess` file. Does that help?

``` apache
<IfModule mod_rewrite.c>
RewriteEngine on
RewriteBase /~sgls22/
DirectoryIndex index.html yellow.php
RewriteRule ^(cache|content|system)/ error [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule ^ yellow.php [L]
</IfModule>
```

**Comment by glasen** [comment]
With a little trial'n error i got the same solution and it works.

Now i've still have two problems:

The site does not use the theme and i can't login into it (user.ini has a username and password)

I've added the debug-statement to core.php so you can maybe see what is going wrong.

**Comment by markseuffert** [comment]
Can you remove the debug statement please and try again?

**Comment by glasen** [comment]
Removed it. Still no theme and no edit.

**Comment by markseuffert** [comment]
Sorry, I don't know why it isn't working on your web server. Your Apache web server returns an error code 403 for many files: You don't have permission to access  /~sgls22/media/images/icon.png on this server. 
Please check the server logfile, contact your university's support if you don't have access to the logfile.

**Comment by glasen** [comment]
Okay, i'll try it.

Strange thing is that yesterday i succesfully installed Wordpress on this account (With Sqlite-plugin) and i worked flawless.

**Comment by msharp1** [comment]
It's seems glasen is giving a try to Grav :)

My general feeling about this issue is that relying on rewrite rules to feed the CMS engine engage the webmaster with many potentials issues with the webserver configuration, that last being:
- Not always alterable by the webmaster
- Not always straight-forward to understand to a webmaster
- Specific to the used webserver

To my experience with several CMS testing around, I always started having issues when coming to optional URLs rewriting for SEO purpose - but in this case it's not vital for basic CMS operation. with Yellow - while I understand the benefits of it - the rewrite rules totally condition if the CMS will be able to operate or not. 

I'm not sure that by saying that it adds something useful to the original issue, sorry about it.

To me, it would be technically more tolerant if Yellow was able operate without rewrite rules. I can't say for sure how much Yellow rely on those... But I can imagine in a perfect world that Yellow would optionally allow the webmaster to enhance Yellow experience and performance by activating the .htaccess layer? Something like that.

**Comment by msharp1** [comment]
> What do you think? If you want to help, please update the Yellow documentation and create new issues for installation problems.

I think it's very welcome. I'm ready to help. Do I have the corresponding permissions on the wiki ? (sorry I'm not that experienced with github as a writer).

**Comment by markseuffert** [comment]
@glasen did we lose you? :)

I think we need more user feedback. It would really good to hear what kind of problems people encounter with web servers. I wasn't aware that there were so many problems with Apache. To be honest I tested Yellow and other flat-file-CMS on different web hostings and things usually worked. Unfortunately, Yellow requires a web server with PHP and rewrite module - so it's not possible to work without it. Maybe it's possible to detect Apache problems automatically and show troubleshooting. Ideally the installation should just work... and in the worst case without reading a lot of documentation. 

@wunderfeyd what do you think?

**Comment by glasen** [comment]
@calimou:
Funny thing, after meddling around with Yellow, i tried Grav. And guess what? It worked OotB. I only had to change RewriteBase

@markseu:
PHP and the rewrite-module are activated on my universities Apache-Installation. As i wrote, Wordpress and now Grav work flawlessly.

**Comment by markseuffert** [comment]
Ouch, then we really have to investigate this.

**Comment by wunderfeyd** [comment]
@markseu: Do we know the exact (or enough of the) configuration of these webservers to reproduce this behaviour locally?

**Comment by glasen** [comment]
I've tried Yellow on my own server. There i have no problems. Everything works perfectly. I don't even have to modify the htaccess-file when i install Yellow to a subfolder.

What do you need from me, so you can reproduce my problems with my university account. The problem is, i have no possibility to login with SSH. I have only an FTP-account and there are also no log files for me :-(

**Comment by glasen** [comment]
Strange thing. I don't know why but know it works.

I only modified rewritebase to "/~sgls22/ and WebinterfaceServerScheme to "https". The same thing i did a few hours ago but there i didn't work.

**Comment by msharp1** [comment]
BTW @glasen; it is a bit off topic but Grav heavily rely on command line... One day or another I'm pretty sure you'll need some way to run CLI command to manage it.

**Comment by wunderfeyd** [comment]
@glasen: Is your site accessed with https? Can you make a backup of your current data and install a fresh Yellow with just these changes to double-check it works when configured this way?

Thank you for the offer to use your university account/site for checking some things, but first let's look whether it works with your changes and a clean Yellow installation.

**Comment by glasen** [comment]
Yes, the site is accessed with https. I tried another clean installation and changing the two options generates a working installation.

**Comment by wunderfeyd** [comment]
Thank you very much for your fast double-check. It's great you found a solution.

@markseu: Do you see any way to detect a wrong RewriteBase and any chance to auto detect the WebinterfaceServerScheme?

**Comment by markseuffert** [comment]
Yellow can detect the rewrite module now, this should catch typical Apache configuration problems.

Yellow can detect the WebinterfaceServerScheme now, this should use HTTPS automatically. Use the [latest plugins](https://github.com/datenstrom/yellow/tree/master/system/plugins) and remove the WebinterfaceServerScheme setting from your `config.ini` file.

Yellow cannot detect the RewriteBase... as far as I know. The rewrite module should detect the location, but it uses the current path under some circumstances. It looks like all CMS suffer from this problem. On a shared web server you need to configure the correct RewriteBase in your `.htaccess` file. In the future someone could create a troubleshooting plugin, with more extensive testing of known problems.

Hope this makes a better installation experience. Please try it out.

**Comment by markseuffert** [comment]
Let's close this issue and see if more problems are reported?

**Comment by wunderfeyd** [comment]
Yup, good way to go.
